cmd/snapshot_export.go

```diff
@@ -17,38 +27,177 @@ func (o *snapshotExportOutput) toJSON()  { outputJSON(o) }
 func (o *snapshotExportOutput) toText()  { outputText(o) }
 func (o *snapshotExportOutput) toTable() { outputTable(o) }
 
-func init() {
-	snapshotCmd.AddCommand(&cobra.Command{
-		Use:   "export <snapshot ID>",
-		Short: "export snapshot",
-		Long: fmt.Sprintf(`This command exports a volume snapshot.
+var snapshotExportCmd = &cobra.Command{
+	Use:   "export <snapshot ID>",
+	Short: "export snapshot",
+	Long: fmt.Sprintf(`This command exports a volume snapshot.
 
 Supported output template annotations: %s`,
-			strings.Join(outputterTemplateAnnotations(&snapshotExportOutput{}), ", ")),
-		RunE: func(cmd *cobra.Command, args []string) error {
-			if len(args) != 1 {
-				return cmd.Usage()
-			}
-			return output(exportSnapshot(args[0]))
-		},
-	})
+		strings.Join(outputterTemplateAnnotations(&snapshotExportOutput{}), ", ")),
+	RunE: func(cmd *cobra.Command, args []string) error {
+		if len(args) != 1 {
+			return cmd.Usage()
+		}
+
+		filePath, err := cmd.Flags().GetString("download")
+		if err != nil {
+			return err
+		}
+
+		snapshot, err := exportSnapshot(args[0])
+		if err != nil {
+			return err
+		}
+
+		if !cmd.Flags().Changed("download") {
+			return output(&snapshotExportOutput{
+				URL:      snapshot.PresignedURL,
+				Checksum: snapshot.MD5sum,
+			}, nil)
+		}
+
+		filePath, err = downloadExportedSnapshot(filePath, args[0], snapshot.PresignedURL)
+		if err != nil {
+			return err
+		}
+
+		fmt.Print("Verifying downloaded file checksum... ")
+
+		if err = checkExportedSnapshot(filePath, snapshot.MD5sum); err != nil {
+			return err
+		}
+
+		fmt.Println("success")
+
+		return nil
+	},
 }
 
-func exportSnapshot(snapshotID string) (outputter, error) {
+func exportSnapshot(snapshotID string) (*egoscale.ExportSnapshotResponse, error) {
 	id, err := egoscale.ParseUUID(snapshotID)
 	if err != nil {
 		return nil, err
 	}
 
-	res, err := asyncRequest(&egoscale.ExportSnapshot{ID: id}, fmt.Sprintf("exporting snapshot %q", id))
+	res, err := asyncRequest(&egoscale.ExportSnapshot{ID: id}, fmt.Sprintf("Exporting snapshot %q", id))
 	if err != nil {
 		return nil, err
 	}
 
-	snapshot := res.(*egoscale.ExportSnapshotResponse)
+	return res.(*egoscale.ExportSnapshotResponse), nil
+}
+
+func downloadExportedSnapshot(filePath, snapshotID, url string) (string, error) {
+	filePath = filepath.ToSlash(filePath)
+
+	st, err := os.Stat(filePath)
+	if err == nil {
+		if st.IsDir() {
+			filePath = filepath.Join(filePath, snapshotID+".qcow2")
+			_, err = os.Stat(filePath)
+			if !os.IsNotExist(err) {
+				return "", fmt.Errorf("file %q: already exists", filePath)
+			}
+		} else {
+			return "", fmt.Errorf("file %q: already exists", filePath)
+		}
+	}
 
-	return &snapshotExportOutput{
-		URL:      snapshot.PresignedURL,
-		Checksum: snapshot.MD5sum,
-	}, nil
+	if err != nil {
```

Comment:
Here if `err` from `st, err := os.Stat(filePath)`  is not nil you should check before if it's not a standard error.
like here: https://github.com/exoscale/cli/blob/master/cmd/sos_download.go#L63-L68